### PR TITLE
Feature/staypoint engine

### LINKED ISSUE #96: Integrate staypoint detection into inference computation
This is the follow up card of the first shot of a staypoint detection integration (see https://github.com/sitcomlab/simport-learning-app/issues/89). After assessing a first feasibility, now an inference engine using the staypoint detection algorithm needs to be implemented (working title: staypoint engine).

- [x] create staypoint engine skeleton (that satisfies inference engine interface)
- [x] create tests with data (possible re-use existing tests of simple engine)
- [x] implement staypiont engine logic for inferences (e.g. adjust scorings etc.)
- [x] simple mechanisms (for now: in code) to switch between engines
- [x] implement convex hull
- [x] make `infer` of inference engine interface async

### PR DESCRIPTION
**Pull request by jbraese**
Infer home/work location based on clustered staypoints

### CODE REVIEW

File: src/app/shared-services/inferences/engine/simple-engine/index.ts [hunk 1]
```diff
@@ -5,11 +5,11 @@ import {
   InferenceDefinition,
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/simple-engine/index.ts]: Why is this file renamed to `index.ts` rather than staying with `simple-engine.ts`?

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import { Inference } from 'src/app/model/inference'
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: Like above: Why `index.ts` rather than `staypoint-engine.ts`?

File: src/app/model/staypoints.ts [hunk 3]
```diff
@@ -8,6 +8,13 @@ export interface StayPoints extends StayPointData {
   trajID: string
```
Comment (schrooom) [inline src/app/model/staypoints.ts:8]: Isn't this model resembling a single `StayPoint` rather than multiple `StayPoints`? If so: this class should be named accordingly, same goes for the filename.

File: src/app/model/staypoints.ts [hunk 4]
```diff
@@ -39,3 +46,36 @@ export function writeStaypointsToGeoJSON(sp: StayPoints, path: string) {
     }
   })
 }
+
+export function writeStaypointClusterArrayToGeoJSON(
+  spcs: StayPointCluster[],
```
Comment (schrooom) [inline src/app/model/staypoints.ts]: You could write this out as `stayPointClusters` instead of `spcs`.

File: src/app/model/staypoints.ts [hunk 5]
```diff
@@ -39,3 +46,36 @@ export function writeStaypointsToGeoJSON(sp: StayPoints, path: string) {
...
+  path: string
+) {
+  const geojson = {
+    name: 'StayPointClusters',
+    type: 'FeatureCollection',
+    features: [],
+    properties: {
+      trajID: spcs[0].trajID,
```
Comment (schrooom) [inline src/app/model/staypoints.ts]: Make sure that `spcs` is not empty before accessing it.

File: src/app/model/staypoints.ts [hunk 6]
```diff
@@ -39,3 +46,36 @@ export function writeStaypointsToGeoJSON(sp: StayPoints, path: string) {
...
+    },
+  }
+  for (const spc of spcs) {
+    geojson.features.push({
+      type: 'Feature',
+      geometry: {
+        type: 'Point',
+        coordinates: [spc.coordinates[1], spc.coordinates[0]],
```
Comment (schrooom) [inline src/app/model/staypoints.ts]: Same here: What happens, if `spc.coordinates.length === 0`?

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+  constructor(private staypointService: StaypointService) {}
+
+  private inputCoordinatesLimit = 100000
+
+  async infer(
+    trajectory: Trajectory,
+    inferences: InferenceDefinition[]
+  ): Promise<InferenceResult> {
+    if (trajectory.coordinates.length > this.inputCoordinatesLimit) {
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: Just a quick note and possibly not highly relevant:
This limitation was necessary with the old `simple-engine`, because we're running DBScan for _all_ locations. In the case of this `staypoint-engine` this _might_ be obsolet (or at least a higher limit could be viable).

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+          ? InferenceResultStatus.noInferencesFound
+          : InferenceResultStatus.successful,
+      inferences: inferenceResults,
+    }
+  }
+
+  private countDays(trajectory: Trajectory): number {
+    // TODO update this method when start/stop are encoded in trajectory
+    const diffMs =
+      trajectory.timestamps[trajectory.timestamps.length - 1].getTime() -
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: I'm not entirely sure: Is it given, that `trajectory.timestamps` is _not_ empty, when `trajectory.coordinates` is? Otherwise you might want to check, if `trajectory.timestamps.length > 0`

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        return inferWorkFromStayPointClusters(
+          stayPointClusters,
+          weekDaysInTrajectory
+        )
+      }
+    })
+    // flatten and filter
+    const inferenceResults: Inference[] = [].concat
+      .apply([], inferenceResultsNested)
+      .filter((i) => i)
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: What is the purpose of this `filter`? If I'm not mistaken, it should be without effect, right?

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      if (i.type === InferenceType.work) {
+        return inferWorkFromStayPointClusters(
+          stayPointClusters,
+          weekDaysInTrajectory
+        )
+      }
+    })
+    // flatten and filter
+    const inferenceResults: Inference[] = [].concat
+      .apply([], inferenceResultsNested)
```
Comment (schrooom) [inline src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: I'm not getting, what you're aiming for with this. Please explain briefly :)

File: src/app/shared-services/inferences/inference.service.ts [hunk 11]
```diff
@@ -29,7 +31,8 @@ export class InferenceFilterConfiguration {
 })
 export class InferenceService implements OnDestroy {
   private static inferenceIntervalMinutes = 240 // 4 hours
-  private inferenceEngine = new SimpleEngine()
+  // private inferenceEngine = new SimpleEngine()
```
Comment (schrooom) [inline src/app/shared-services/inferences/inference.service.ts]: Maybe you could add a flag that determines the used engine and remove this commented code? (Ideally we even would some sort of `InferenceEngineFactory`, that simply returns an `InferenceEngine` to this service – `InferenceService` doesn't need to know, what kind of engine in particular it gets.)

File: src/app/shared-services/staypoint/staypoint-clusterer.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+   * @param  neighborhoodRadius The size of the search radius from a given point within which points are classified as neighbors.
+   * @param  pointsInNeighborhood Minimum number of points required to form a neighborhood ('dense region').
+   * @return The clusters of staypoints
+   */
+  public clusterStayPoints(
+    stayPoints: StayPoints,
+    neighborhoodRadius: number,
+    pointsInNeighborhood: number
+  ): StayPointCluster[] {
+    if (stayPoints.coordinates.length === 0) return undefined
```
Comment (schrooom) [inline src/app/shared-services/staypoint/staypoint-clusterer.ts:22]: Depending on what you want to go for with your implementation, you could also return `[]`.

File: src/app/shared-services/staypoint/utils/index.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+import concaveman from 'concaveman'
+
+/**
+ * Get the location of the three clusters of staypoints where the most nights (12pm to 4am) were spent.
+ * @param stayPointClusters The clusters of staypoints from which home location is to be extracted
+ * @param daysInTrajectory The number of days for which the trajectory has been recorded,
+ * used to calculate the "confidence" value (number of nights spent at staypoint/number of days in trajectory)
+ * @return The top three home inferences
+ */
+export function inferHomeFromStayPointClusters(
```
Comment (schrooom) [inline src/app/shared-services/staypoint/utils/index.ts]: This seems a rather powerful `utils`-file and a single point, where a lot of the `staypoint-engine`'s function lays.
Therefore I would expect some of these methods in the `staypoint-engine` itself (e.g. `inferHomeFromStayPointClusters(..)`) – especially, if they are exclusively used over there.

**Comment by schrooom** [comment]
Looks good overall, nicely done! :) I commented on a few smaller things – don't hesitate to ask any question you might have ;)

(I've yet to look at the `spec.ts`-files)

File: src/app/shared-services/inferences/engine/simple-engine/index.ts (hunk 1 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/simple-engine/index.ts]: agreed, I renamed it.

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts (hunk 2 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: same as above, I renamed it.

File: src/app/model/staypoints.ts (hunk 3 above)
Comment (jbraese) [reply src/app/model/staypoints.ts:8]: This structure contains all staypoints for a trajectory, thus I'd leave the name as is.

File: src/app/model/staypoints.ts (hunk 4 above)
Comment (jbraese) [reply src/app/model/staypoints.ts]: Agreed, I changed it.

File: src/app/model/staypoints.ts (hunk 5 above)
Comment (jbraese) [reply src/app/model/staypoints.ts]: Makes sense, I changed it.

File: src/app/model/staypoints.ts (hunk 6 above)
Comment (jbraese) [reply src/app/model/staypoints.ts]: A staypoint cluster is required to have exactly two coordinates, so I think this is fine.

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts (hunk 7 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: This is a good point and I agree it might be obsolete.

I'd not change it for now due to the issue with the test trajectory we talked about once: the nepal test trajectory is not saved in the database and hence there are issues with ID as the foreign key when trying to save its inferences to the database. I believe the simple-engine is only avoiding running into this issue through this check, so I am leaving it in here for now as well. At some point it would probably make sense to address this edge case with the test trajectory type, I suggest removing this check after.

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts (hunk 8 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: I think this is fine as is, it would be highly problematic in a number of places if the length of coordinates and timestamps in a trajectory don't match (ie if somehow GPS points without timestamps or vice versa are saved), and I haven't seen it occur so far.

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts (hunk 9 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: this filters "falsey" values, like if there are no or undefined inferences.

File: src/app/shared-services/inferences/engine/staypoint-engine/index.ts (hunk 10 above)
Comment (jbraese) [reply src/app/shared-services/inferences/engine/staypoint-engine/index.ts]: this flattens `[[homeInference1, homeInference2, homeInference3], [workInference1, workInference2, workInference3]]` into `[homeInference1, homeInference2, homeInference3, workInference1, workInference2, workInference3]`

File: src/app/shared-services/inferences/inference.service.ts (hunk 11 above)
Comment (jbraese) [reply src/app/shared-services/inferences/inference.service.ts]: Agreed, I went with the flag option for now to not introduce too much additional complexity into this already quite elaborate inference engine architecture.

File: src/app/shared-services/staypoint/staypoint-clusterer.ts (hunk 12 above)
Comment (jbraese) [reply src/app/shared-services/staypoint/staypoint-clusterer.ts:22]: To keep it consistent with the rest of the staypoint modules I would leave it at undefined.

File: src/app/shared-services/staypoint/utils/index.ts (hunk 13 above)
Comment (jbraese) [reply src/app/shared-services/staypoint/utils/index.ts]: Agreed, I pulled it into `staypoint-engine`.

**Review by schrooom — APPROVED** [review]
